test/unit/Cardano/Wallet/DB/MVarSpec.hs

```diff
@@ -152,52 +188,95 @@ dbMergeTxHistoryProp db (KeyValPairs keyValPairs) =
-    prop = monadicIO $ liftIO $ do
-        forM_ keyValPairs $ \(key, val) -> do
-            putCheckpoint db key (initWallet $ DummyState 0)
-            unsafeRunExceptT $ putTxHistory db key val
-        forM_ keyValPairs $ \(key, _) -> do
-            res <- readTxHistory db key
-            res `shouldBe` (Map.unions (snd <$> restrictTo key keyValPairs))
-
-dbPutTxHistoryBeforeCheckpointProp
+    prop = monadicIO $ do
+        cp <- pick arbitrary
+        meta <- pick arbitrary
+        liftIO $ do
+            unsafeRunExceptT $ once keyValPairs $ \(k, _) -> createWallet db k cp meta
+            unsafeRunExceptT $ forM_ keyValPairs $ uncurry (putTxHistory db)
+            forM_ keyValPairs $ \(key, _) -> do
+                res <- readTxHistory db key
+                res `shouldBe` (Map.unions (snd <$> restrictTo key keyValPairs))
+
+dbPutTxHistoryBeforeWalletProp
     :: DBLayer IO DummyState
     -> PrimaryKey WalletId
     -> Property
-dbPutTxHistoryBeforeCheckpointProp db key@(PrimaryKey wid) = monadicIO $ liftIO $ do
+dbPutTxHistoryBeforeWalletProp db key@(PrimaryKey wid) = monadicIO $ liftIO $ do
     runExceptT (putTxHistory db key mempty) >>= \case
         Right _ -> fail "expected insertion to fail but it succeeded?"
         Left err -> err `shouldBe` ErrNoSuchWallet wid
     readTxHistory db key `shouldReturn` mempty
 
-dbPutTxHistoryNoEffectOnWallet
+dbPutTxHistoryIsolationProp
     :: DBLayer IO DummyState
-    -> (PrimaryKey WalletId, DummyState, Map (Hash "Tx") (Tx, TxMeta))
+    -> PrimaryKey WalletId
+    -> (Wallet DummyState, WalletMetadata, Map (Hash "Tx") (Tx, TxMeta))
+    -> Map (Hash "Tx") (Tx, TxMeta)
     -> Property
-dbPutTxHistoryNoEffectOnWallet db (key, s, txs) = monadicIO $ liftIO $ do
-    let cp = initWallet s
-    putCheckpoint db key cp
-    void $ runExceptT (putTxHistory db key txs)
-    cp' <- readCheckpoint db key
-    cp' `shouldBe` Just cp
-
-dbPutCheckpointNoEffectOnHistory
+dbPutTxHistoryIsolationProp db key (cp,meta,txs) txs' = monadicIO $ liftIO $ do
+    unsafeRunExceptT $ createWallet db key cp meta
+    unsafeRunExceptT $ putTxHistory db key txs
+    unsafeRunExceptT $ putTxHistory db key txs'
+    readCheckpoint db key `shouldReturn` Just cp
+    readWalletMeta db key `shouldReturn` Just meta
+
+dbPutCheckpointIsolationProp
     :: DBLayer IO DummyState
-    -> (PrimaryKey WalletId, Map (Hash "Tx") (Tx, TxMeta))
+    -> PrimaryKey WalletId
+    -> (Wallet DummyState, WalletMetadata, Map (Hash "Tx") (Tx, TxMeta))
+    -> Wallet DummyState
     -> Property
-dbPutCheckpointNoEffectOnHistory db (key, txs) = monadicIO $ liftIO $ do
-    let cp0 = initWallet (DummyState 0)
-    let cp1 = initWallet (DummyState 14)
-    putCheckpoint db key cp0
-    void $ runExceptT (putTxHistory db key txs)
-    putCheckpoint db key cp1
-    txs' <- readTxHistory db key
-    txs' `shouldBe` txs
+dbPutCheckpointIsolationProp db key (cp,meta,txs) cp' = monadicIO $ liftIO $ do
```

Comment:
there is a number of tests that are permutations of `put{TxHistory, WalletMeta, Checkpoint}`. Basically, we have : 
1. CreateWallet
2. Two puts
3. Two reads where one does not overlap with one of 2.
I wonder if we can replace all those tests with one, where we create wallet, than **shuffle** list of pairs (operation and arg like txs, cp meta). Moreover, we can make it more general. For instance, generate :  
``` 
([Wallet DummyState], [WalletMetadata], [Map (Hash "Tx") (Tx, TxMeta)])
```
and even make sure the lengths of those three lists are different. then construct operations with generated values using those. Then shuffle. In that case we will test a number of what should be independent operations, some of them updating and at the end we expect **last** values of each list to be returned upon corresponding reads ...

Of course this can be a separate PR - but I think it would be valuable to have those tests, mainly because we will end up with one sequential test and one parallel test for a lot of permutations and multiple updates

Reply:
Indeed! That's what I mentioned to you on Slack. I am not yet fully satisfied with the testing situation here, although it tests things, we can feel the duplication and weight of the tests here.
Although it's tricky to come up with a unified approach for operation agnostic properties here, i think it's valuable in the long run.